File: ffi/python/blsct/out_point.py
from . import blsct
from .managed_obj import ManagedObj
from .serializable import Serializable
from .ctx_id import CtxId
from typing import Any, override, Self, Type

class OutPoint(ManagedObj, Serializable):
  """
  Represents an outpoint of a confidential transaction. Also known as `COutPoint` on the C++ side.

  >>> from blsct import OutPoint, TxId, TX_ID_SIZE
  >>> import secrets
  >>> tx_id = TxId.from_hex(secrets.token_hex(TX_ID_SIZE))
  >>> out_index = 0
  >>> OutPoint.generate(tx_id, out_index)
  OutPoint(<Swig Object of type 'void *' at 0x105b071b0>)  # doctest: +SKIP
  """
  @classmethod
  def generate(cls: Type[Self], ctx_id: CtxId, out_index: int) -> Self:
    """Generate an outpoint from a transaction ID and output index."""
    rv = blsct.gen_out_point(ctx_id.serialize(), out_index)
    inst = cls(rv.value)
    blsct.free_obj(rv)
    return inst

  def serialize(self) -> str:
    """Serialize the OutPoint to a hexadecimal string"""
    return blsct.serialize_out_point(self.value())

  @classmethod
  @override
  def deserialize(cls, hex: str) -> Self:
    """Deserialize the OutPoint from a hexadecimal string"""
    return blsct.deserialize_out_point(hex)

  @override
  def value(self) -> Any:
    return blsct.cast_to_out_point(self.obj)

  @classmethod
  def default_obj(cls) -> Any:
    raise NotImplementedError("Cannot create an OutPoint without required parameters.")

File: ffi/ts/src/outPoint.ts
import {
  castToOutPoint,
  deserializeOutPoint,
  freeObj,
  genOutPoint,
  serializeOutPoint,
} from './blsct'

import { ManagedObj } from './managedObj'
import { CTxId } from './ctxId'

/** Represents an outpoint of a confidential transaction. Also known as `COutPoint` on the C++ side.
  *
  * Examples:
  * ```ts
  * const { OutPoint, CTxId, CTX_ID_SIZE } = require('navio-blsct')
  * const { randomBytes } = require('crypto')
  * const cTxIdHex = randomBytes(CTX_ID_SIZE).toString('hex')
  * const cTxId = CTxId.deserialize(cTxIdHex)
  * const outPoint = OutPoint.generate(cTxId)
  * outPoint
  * const ser = outPoint.serialize()
  * const deser = OutPoint.deserialize(ser)
  * ser === deser.serialize() // true
  * ```
  */
export class OutPoint extends ManagedObj {
  constructor(obj: any) {
    super(obj)
  }
  
  /** Generates a new `OutPoint` instance.
   *
   * @param ctxId - The transaction ID of the outpoint.
   * @return A new `OutPoint` instance with the specified transaction ID.
   */
  static generate(
    ctxId: CTxId,
  ) {
    const rv = genOutPoint(ctxId.serialize())
    const obj = rv.value
    freeObj(rv)
    return new OutPoint(obj)
  }

  override value(): any {
    return castToOutPoint(this.obj)
  }

  override serialize(): string {
    return serializeOutPoint(this.value())
  }

  /** Deserializes an `OutPoint` from its hexadecimal representation.
   *
   * @param hex - The hexadecimal string to deserialize.
   * @returns A new `OutPoint` instance.
   */
  static deserialize(
    this: new (obj: any) => OutPoint,
    hex: string
  ): OutPoint {
    return OutPoint._deserialize(hex, deserializeOutPoint)
  }
}

